sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/batch/implementation/BatchAsyncContainerExecutor.java

```diff
@@ -0,0 +1,286 @@
+            .flatMap(collection -> {
+                PartitionKeyDefinition definition = collection.getPartitionKey();
+                PartitionKeyInternal partitionKeyInternal = getPartitionKeyInternal(operation, definition);
+                operation.setPartitionKeyJson(partitionKeyInternal.toJson());
+
+                return this.docClientWrapper.getPartitionKeyRangeCache()
+                    .tryLookupAsync(BridgeInternal.getMetaDataDiagnosticContext(null), collection.getResourceId(), null, null)
+                    .map((Utils.ValueHolder<CollectionRoutingMap> routingMap) -> {
+
+                        return routingMap.v.getRangeByEffectivePartitionKey(
+                            getEffectivePartitionKeyString(
+                                partitionKeyInternal,
+                                definition)).getId();
+                    });
+        });
+
+        return pkRangeIdMono;
+    }
+
+    private static void validateOperationEpk(
+        final ItemBatchOperation<?> operation,
+        final RequestOptions options) {
+
+        checkNotNull(operation, "expected non-null operation");
+        checkNotNull(options, "expected non-null options");
+
+        final Map<String, Object> properties = options.getProperties();
+        if (properties == null) {
+            return;
+        }
+
+        if (properties.containsKey(BackendHeaders.EFFECTIVE_PARTITION_KEY)
+            || properties.containsKey(BackendHeaders.EFFECTIVE_PARTITION_KEY_STRING)
+            || properties.containsKey(HttpHeaders.PARTITION_KEY)) {
+
+            byte[] epk = (byte[]) properties.computeIfPresent(BackendHeaders.EFFECTIVE_PARTITION_KEY,
+                (k, v) -> v instanceof byte[] ? v : null);
+
+            String epkString = (String) properties.computeIfPresent(BackendHeaders.EFFECTIVE_PARTITION_KEY_STRING,
+                (k, v) -> v.getClass()  == String.class ? v : null);
+
+            String pkString = (String) properties.computeIfPresent(HttpHeaders.PARTITION_KEY,
+                (k, v) -> v.getClass() == String.class ? v : null);
+
+            checkState(!((epk == null && pkString == null) || (epkString == null)),
+                "expected byte array value for {0} and string value for {1} when either property is set",
+                BackendHeaders.EFFECTIVE_PARTITION_KEY,
+                BackendHeaders.EFFECTIVE_PARTITION_KEY_STRING);
+
+            checkState(operation.getPartitionKey() == null,
+                "partition key and effective partition key may not both be set.");
+        }
+    }
+
+    private Semaphore getOrAddLimiterForPartitionKeyRange(String partitionKeyRangeId) {
+        return this.limiters.computeIfAbsent(partitionKeyRangeId, id -> new Semaphore(1));
+    }
+
+    private BatchAsyncStreamer getOrAddStreamerForPartitionKeyRange(String partitionKeyRangeId) {
+
+        return this.streamers.computeIfAbsent(partitionKeyRangeId, id -> new BatchAsyncStreamer(
```

Comment:
Is this making sure only **1 Streamer exists for each PKRangeId**? Reference on C# https://github.com/Azure/azure-cosmos-dotnet-v3/blob/master/Microsoft.Azure.Cosmos/src/Batch/BatchAsyncContainerExecutor.cs#L258

Reply:
Are we testing this under concurrent operations to verify there is only 1 Streamer?

Reply:
Haven't tested it with code, but Java doc says that function will be executed at most once per key: [doc](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ConcurrentHashMap.html#computeIfAbsent-K-java.util.function.Function-). So should be fine.

Reply:
Have verified now with adding logs on streamer init.